### DISCUSSION TITLE
Help using GEM with Adafruit ESP32-S3 Reverse TFT

### DISCUSSION THREAD

**Discussion by davexre** [Spirik/GEM#99] [Q&A] [answered]
I'm hoping to get GEM up and running with this little guy: https://learn.adafruit.com/esp32-s3-reverse-tft-feather - it uses an ST7789 TFT controller, which is a little different from the ST7735 a lot of the Adafruit examples were built on, but it's supported by Adafruit GFX, and I'm kind of at a loss as to what's missing (without tearing apart GEM's Adafruit support looking for things that might somehow be incompatible. 

I initially tried implementing a POC into my project, but I'd end up back at the UF2 bootloader, basically exactly as described in this thread: https://forums.adafruit.com/viewtopic.php?t=205556 - the root cause there was apparently old library versions, but I'm on current (as of today) libraries for everything, as near as I can tell. 

So, I fell back to trying the GEM blink example, with the required changes for my display (code below). I never get to the Serial.println statements I added - after I reset the board (required after uploading a sketch), the builtin LED just fast blinks, and that's it. In contrast, running Adafruit's "Blink" and GFX example test code works as expected. I'm guessing some incompatibility - but figured it would be worth asking here before I try to dive into the GEM code, and also see if there are any pointers to sorting this out?

Thanks!

Here's the code I changed in the provided Blink.cpp

```
...
...
#include <Adafruit_ST7789.h>
#include <Arduino.h>
#include <GEM_adafruit_gfx.h>
#include <KeyDetector.h>
...
...
// Changed pins to match my config - this uses the three builtin buttons, and three external

// Pins the buttons are connected to
const byte downPin = 2;
const byte leftPin = 9;
const byte rightPin = 6;
const byte upPin = 5;
const byte cancelPin = 1;
const byte okPin = 0;
...
...

Adafruit_ST7789 tft = Adafruit_ST7789(TFT_CS, TFT_DC, TFT_RST);

...
...

void setup() {
  // Push-buttons pin modes
  pinMode(downPin, INPUT_PULLDOWN);
  pinMode(leftPin, INPUT_PULLDOWN);
  pinMode(rightPin, INPUT_PULLDOWN);
  pinMode(upPin, INPUT_PULLDOWN);
  pinMode(cancelPin, INPUT_PULLDOWN);
  pinMode(okPin, INPUT_PULLUP);

  // Serial communications setup
  Serial.begin(115200);

  Serial.println("Hello world!");

  // With the Adafruit ESP32-S3 Reverse TFT, we have to turn on the backlite and
  // power to the display itself...
  pinMode(TFT_BACKLITE, OUTPUT);
  digitalWrite(TFT_BACKLITE, HIGH);

  Serial.println("TFT backlite on");

  // turn on the TFT / I2C power supply
  pinMode(TFT_I2C_POWER, OUTPUT);
  digitalWrite(TFT_I2C_POWER, HIGH);
  delay(10);

  Serial.println("TFT turned on");

  // Use this initializer if using a 1.8" TFT screen:
  //   tft.initR(INITR_BLACKTAB);  // Init ST7735S chip, black tab
  // OR use this initializer if using a 1.8" TFT screen with offset such as
  // WaveShare: tft.initR(INITR_GREENTAB);   // Init ST7735S chip, green tab See
  // more options in Adafruit GFX library documentation

  // Optionally, rotate display
  // tft.setRotation(3); // See Adafruit GFX library documentation for details

  // no tft.initR() call in the ST7789 library - just init, instead
  tft.init(240, 135);  // Init ST7789 240x135
  tft.setRotation(3);

  // Menu init, setup and draw
  menu.init();
  setupMenu();
  menu.drawMenu();
}
```

**Comment by Spirik** [comment]
Hello, @davexre !
I'd suggest to try [Basic](https://github.com/Spirik/GEM/wiki/Example-01%3A-Basic) example or to create barebone sketch with GEM and continue investigation from there, e.g. create single menu page with single menu item (or even without menu items completely). Even try version w/o KeyDetector (and even better, try KeyDetector [example](https://github.com/Spirik/KeyDetector/wiki/Example-02%3A-Digital-signal-readings) sketch alone w/o GEM). Maybe this is somehow a memory issue (however from what I can tell, this board should have plenty).

**Reply by davexre** [reply]
@Spirik - thanks for the ideas! Yeah, cut the problem down - makes sense. I'll dig into this as I have time this week, and let you know how it progresses!

I agree that it shouldn't be memory - I'm using something like 20% of flash, and < 10% of RAM, based on the compilation size numbers... But weirder things have happened, so who knows... LOL

**Comment by Spirik** [comment]
@davexre Also, explicitly setting background and foreground colors may help with color-capable display issues, like it was done in [this version](https://gist.github.com/Spirik/5e3ce403a49f42c1c066989b0342b657#file-example-01_basic_feather-ino-L105) of Basic example adapted for M0-based Feather. However that won't affect an issue with missing output to Serial, which you currently have.

**Comment by dave-re-imprivata** [comment]
Hmm... this might actually be a PlatformIO project issue... I completely replaced the code in the project with Adafruit's demo code, and it does the same thing. I'm going to start over with a fresh project, and see what develops from there!

**Reply by Spirik** [reply, marked as answer]
Oh, interesting. Probably, that is may be the case indeed, given that PlatformIO currently have struggles with support for most recent versions of ESP32 core. See this [thread](https://github.com/platformio/platform-espressif32/issues/1225).

**Comment by davexre** [comment]
Man, that thread is an adventure. I played with the [pioarduino fork of the Espressif platformio support](https://github.com/pioarduino/platform-espressif32), but that had issues, too. Went back to Arduino IDE, and had some forward progress. I got the basic [Adafruit Blink sketch](https://github.com/Spirik/GEM/blob/master/examples/AdafruitGFX/Example-02_Blink/Example-02_Blink.ino) to compile and load, and it showed the splash screen, but then nothing. KeyDetector was working, however, as witnessed by the serial monitor. The screen with the splash - it showed in a black square in the middle of the (very wide aspect) screen - that square stayed black after the splash stopped, and it showed colored static in the areas outside the square.

So... I cut it way back. First, I switched to the Feather M0 code you posted, and udpated it to match the ST7789 screen needs. Then I cut out all the key detection stuff, and have no menu items. Code is below. I can see that it powers on the back light - and it goes through all the steps in the setup() (judging by the serial output), but all I get is a black screen. I originally had the menu items included, but it was the same story. 

In theory... this should work (the original blink example is ST7735, which should basically work the same as ST7789), but maybe there's an incompatibility between them, or something additional missing. So, that's my next step - start digging into those, and how GEM is using them. Unless you have a better idea 😄 

```
/*
  GEM testing on Adafruit ESP32-S3
*/

#include <GEM_adafruit_gfx.h>
#include <SPI.h>
#include <Wire.h>
#include <Adafruit_ST7789.h>


Adafruit_ST7789 display = Adafruit_ST7789(TFT_CS, TFT_DC, TFT_RST);

// Create variables that will be editable through the menu and assign them initial values
int number = -512;
boolean enablePrint = false;

// Create two menu item objects of class GEMItem, linked to number and enablePrint variables 
// GEMItem menuItemInt("Number:", number);
// GEMItem menuItemBool("Enable print:", enablePrint);

// Create menu button that will trigger printData() function. It will print value of our number variable
// to Serial monitor if enablePrint is true. We will write (define) this function later. However, we should
// forward-declare it in order to pass to GEMItem constructor
// void printData(); // Forward declaration
// GEMItem menuItemButton("Print", printData);

// Create menu page object of class GEMPage. Menu page holds menu items (GEMItem) and represents menu level.
// Menu can have multiple menu pages (linked to each other) with multiple menu items each
GEMPage menuPageMain("Main Menu");

// Create menu object of class GEM_adafruit_gfx. Supply its constructor with reference to display object we created earlier
GEM_adafruit_gfx menu(display, GEM_POINTER_ROW, GEM_ITEMS_COUNT_AUTO);

void setupMenu();

void setup() {

  pinMode(2, INPUT_PULLDOWN);
  pinMode(1, INPUT_PULLDOWN);
  pinMode(0, INPUT_PULLUP);

  Serial.begin(115200);
  while (!Serial) {}

  pinMode(TFT_BACKLITE, OUTPUT);
  digitalWrite(TFT_BACKLITE, HIGH);
  pinMode(TFT_I2C_POWER, OUTPUT);
  digitalWrite(TFT_I2C_POWER, HIGH);

  Serial.println("display and I2C powered on");

  delay(250); 
  display.setRotation(1);

  // Explicitly set correct colors for monochrome OLED screen
  menu.setForegroundColor(ST77XX_WHITE);
  menu.setBackgroundColor(ST77XX_BLACK);

  // Disable GEM splash (it won't be visible on the screen of buffer-equiped displays such as this one any way)
  menu.setSplashDelay(0);
  
  // Menu init, setup and draw

  Serial.println("Calling menu.init()");

  menu.init();

  Serial.println("Calling setupMenu()");

  setupMenu();

  Serial.println("Calling menu.drawMenu()");

  menu.drawMenu();

}

void setupMenu() {
  // Add menu items to menu page
  // menuPageMain.addMenuItem(menuItemInt);
  // menuPageMain.addMenuItem(menuItemBool);
  // menuPageMain.addMenuItem(menuItemButton);

  menu.setMenuPageCurrent(menuPageMain);
}

void loop() {
  delay(2000);
}
```

**Comment by Spirik** [comment]
Thought I'd give it a try on Wokwi.

There is ESP32-S3 board there, however no ST7789-based display. But they have ILI9341 instead, which should be close enough (at least it works with `Adafruit_ST7789.h` library just the same). So here is Wokwi [project](https://wokwi.com/projects/380662202976325633) I've initially found with ESP32-C3 board and ILI9341 running Adafruit demo sketch. It works just fine.

However, when I've made a [copy](https://wokwi.com/projects/404945985875680257) of that project and swapped ESP32-C3 board with ESP32-S3, it stopped to display anything, although Serial works fine. Just like in your experiments.

To be exact, I have no idea what to make of it=) Does S3 that much differ from C3? What should be changed in code to make it work? I used the same pins as where used in initial C3 example. Is this somehow related to SPI?

I will try to dig deeper, but thought I'd let you know my findings so far.

**Reply by davexre** [reply]
Wokwi is super cool - wasn't aware of that site! It's possible the SPI pins are different, yeah. I can dig on that a little after I get done with work today. That'd be my first guess, too. But it's strange that it works with one board, and not the other. That's the Adafruit demo code, there, and *that* works fine on my board here. So, likely we've got a problem with the simulation to work out first.

**Reply by davexre** [reply]
Working: https://wokwi.com/projects/404957370389903361 - just had to get SPI CLK and MOSI wired correctly.

Looks like starting up a sketch, and then pausing it will cause Wokwi to put labels on all the board pins, so you can see how they're being used... In this case, the S3 board has MOSI on pin 11, and CLK on pin 12.

**Reply by Spirik** [reply]
Good catch! So that means GEM on Wokwi version of ESP32-S3 [works as well](https://wokwi.com/projects/404964529727364097).

**Reply by davexre** [reply]
Ok, that's good... now to see why it doesn't want to work on my hardware...!!!

**Comment by davexre** [comment]
Ok, I found my final issue.... I was failing to actually init the display (you can see that in my last code above). Adding:

`display.init(135, 240); `

Turns the display on, and I get the menu now! So, I think the root cause of my initial issue is platformio - thanks again for those pointers. I'll monitor what's going on there, and see if I can get it working again later, but for now... the suboptimal Arduino IDE will have to do, I suppose!

**Reply by Spirik** [reply]
That's great news!

Btw, you can also look at [Arduino Extension](https://marketplace.visualstudio.com/items?itemName=vsciot-vscode.vscode-arduino) for VS Code.
